### ISSUE TITLE
Several vertical-stack-in-card in a single view are stacked vertically

### ISSUE THREAD

**Issue by pplucky** [ofekashery/vertical-stack-in-card#80]
https://community.home-assistant.io/t/stack-in-card-drop-in-replacement-for-vertical-stack-in-card/180072/51

**Comment by ofekashery** [comment]
I think you meant to create the issue on the [``custom-cards/stack-in-card``](https://github.com/custom-cards/stack-in-card) repo.

**Comment by pplucky** [comment]
Hello.

No actually I meant that the same issue exists for both cards.

Is your card going to be discontinued in favour of stack-in-card?

**Comment by bat86** [comment]
Same problem after upgrade on 3.x version

**Comment by SAOPP** [comment]
Hi guys! Just updated my plugin to the latest 0.3.2 version. I have the same result. Seems like a bug.

**Comment by pplucky** [comment]
It seems to be a [frontend issue](https://github.com/home-assistant/frontend/issues/5321), rather a card one.

**Comment by thecode** [comment]
I can confirm and easily reproduce this.
HA 107.x, with vertical-stack-in-card version 0.2.1 cards are ordered correctly.
With 0.3.x several cards are stacked vertically.

**Comment by SAOPP** [comment]
Eexactly right!

**Comment by blamaz** [comment]
Same here

**Comment by stave1989** [comment]
Is there a solution to this problem ?

**Comment by ofekashery** [comment]
@stave1989 As a temporary solution, you can try to use the [layout-card](https://github.com/thomasloven/lovelace-layout-card).

**Comment by thecode** [comment]
> Is there a solution to this problem ?

You can downgrade to version 0.2.1 for now

**Comment by stave1989** [comment]
@thecode The 0.2.1  "vertical-stack-in-card.js" file is unfortunately not available any more :(

**Comment by krkeegan** [comment]
This is source repository, all versions are always available.

Here is v0.2.1 if you want that, you will have to manually install it.

https://raw.githubusercontent.com/ofekashery/vertical-stack-in-card/v0.2.1/vertical-stack-in-card.js

Alternatively, within the HACS integration control panel. Click on the gear in the top right and change the Number of releases to show to 10 or something until you can see 0.2.1.

**Comment by stave1989** [comment]
Thank you @krkeegan 
I always manually install ;)

**Comment by magiva** [comment]
> > Is there a solution to this problem ?
> 
> You can downgrade to version 0.2.1 for now

downgrading to any available version (from home assistant) does not fix this

**Comment by scstraus** [comment]
I'm also having this problem.

With vertical-stack-in card:

![image](https://user-images.githubusercontent.com/7644023/79504424-29ba8b00-8033-11ea-9501-e9ba8c4d37bc.png)


Without vertical-stack-in card:

![image](https://user-images.githubusercontent.com/7644023/79504078-89fcfd00-8032-11ea-8e37-176fb52a54fb.png)

**Comment by magiva** [comment]
I've had to stop using it because of this. only workaround for me is to use
the standard v card

br
Mark-Andrew

**Comment by scstraus** [comment]
Update: downgrade to 0.2.1 worked for me.

**Comment by magiva** [comment]
thats great im sure but it didnt work for me. repeating it doesnt make it
work ya know :)

**Comment by scstraus** [comment]
Did you clear your cache and click reload in the upper right hand menu of lovelace?

**Comment by NvrBst** [comment]
Can also confirm that the cards are all in a single line when loaded. But if you then Hold-Shift while pressing "Reload" in Chrome, then the view is displayed properly.

But if you exit Lovelace and go back, the view will return to the single line stacked. Until u reload and wipe cache again.

**Comment by hcoohb** [comment]
Same problem here!
It seems to happen to me only if I nest some vertical-stack-in-card though

**Comment by Nicras** [comment]
Same here, wondered why my cards are stacked vertically. Any news on this?

**Comment by scstraus** [comment]
Looks like fix should be in next major release hopefully.

**Comment by krkeegan** [comment]
I fixed the glitch based on the guidance from https://developers.home-assistant.io/blog/2020/06/01/getcardsize/ . However, I am not a javascript developer, and it took me a few hours to decipher the concept of a Promise.  The patch works well for me on version 0.110.4 of HomeAssistant.  I believe you will need at least version 0.110 for it to work.

Hope this helps.  Feel free to clean up my code or abandon it if I have done this in the wrong way.

**Comment by krkeegan** [comment]
I can't stop tinkering.

This is adapted from how the native Horizontal Stack does it:
```javascript
  async getCardSize() {
    const promises = [];
    for (const element of this._refCards) {
      promises.push(computeCardSize(element));
    }

    const results = await Promise.all(promises);

    return results.reduce((a,b) => a + b, 0);
  }
```

https://github.com/home-assistant/frontend/blob/7fb852893ccb956ddd7cf25c70c4ae9ddddab0b9/src/panels/lovelace/cards/hui-horizontal-stack-card.ts

**Comment by NvrBst** [comment]
Yea, and it's officially supported in 0.111 now too:
  https://developers.home-assistant.io/blog/2020/06/01/getcardsize/

**Comment by scstraus** [comment]
So I just upgraded to .111.2 and vertical stack in card 0.3.3, but now I get something a little better than the full error, but not as good as having it fully working right. This is happening on any view where I use the vertical stack in card but not others where I don't. 



![image](https://user-images.githubusercontent.com/7644023/84579054-027e0100-adcb-11ea-80e0-8c65db2d032b.png)

**Comment by scstraus** [comment]
Bummer. Good to know that it's not only me though. I hope someone can pull it off.

**Comment by krkeegan** [comment]
So one thing that works for me, but is a horrible hack is to change getCardSize to this:

```javascript
  getCardSize() {
    return 5;
  }
```

It is a horrible hack as it makes hard codes the card size.  But it at least works for me in the interim.

**Comment by scstraus** [comment]
Wouldn't that make every card the same size though? As I've always had some problems with card size being reported properly, I've had some cards where I use card mod to hardcode the size thus:

```

  - type: custom:mod-card
    report_size: 4
    card:
      type: custom:vertical-stack-in-card
      cards:
<your usual vertical stack cards config here>
```

But it doesn't always work well. If I get desperate, I will go in and measure each card and try to hardcode them in this way.. I'm hoping it doesn't come to that, though.

**Comment by ofekashery** [comment]
@scstraus could you share your ``vertical-stack-in-card`` config of the screenshot you sent above please? I can't reproduce this issue anymore with my configs..

**Comment by scstraus** [comment]
Happily. My whole [config is on my github](https://github.com/scstraus/home-assistant-config). The view I posted the screenshot of is [here](https://github.com/scstraus/home-assistant-config/blob/master/lovelace/Tab3_Heat.yaml) and is described [here](https://github.com/scstraus/home-assistant-config#tab-3-heat). I have other tabs with the same problem, I can send you screenshots of how those are rendering and the config is there too.

And I will gladly test anything you need me to. I am a very heavy user of your card so this particular problem is a major one for me, so anything I can do to make things easier for you to test, please just let me know. We could open up a discord and you can try changes and I can send you the results in real time if you want.

**Comment by scstraus** [comment]
Worth noting that it probably needs to be tested with more than 20 fairly complex cards to reproduce, as it appears to place the first ~20 into ~10 stacks correctly before giving up and throwing the rest into a big column.

**Comment by scstraus** [comment]
Interestingly, I can get the cards to stack correctly if I first open in a 2 column view and then expand window to a 3 column view:

![image](https://user-images.githubusercontent.com/7644023/87035566-d139fa80-c1e9-11ea-93cb-ebd077ab9ee7.png)


![image](https://user-images.githubusercontent.com/7644023/87035411-9637c700-c1e9-11ea-8309-4461f54b2d9c.png)

**Comment by scstraus** [comment]
What do you think about adding an option to the card that allows us just to hardcode the size of the stack and so allowing it to answer with the size rather than to wait for it to be calculated, which in my case appears to lead to timeouts before the "promise" can be returned, likely because I'm using too complex views? I've tried in the past to do this with browser mod but it doesn't really play nicely with stacks.. Seems like this is the simplest solution to this problem.

**Comment by scstraus** [comment]
I ended up moving to [Canary Card](https://github.com/jcwillox/lovelace-canary) which solves this problem for me (finally!).  I also found out that part of my issue was caused with a similar problem with [State Switch Card](https://github.com/thomasloven/lovelace-state-switch/issues/40)

**Comment by ryanwinter** [comment]
Also switched to [Canary Card](https://github.com/jcwillox/lovelace-canary) which solved my this problem for me.

**Comment by pplucky** [comment]
Also ended up moving to [Canary Card](https://github.com/jcwillox/lovelace-canary) without looking back.

**Comment by ofekashery** [comment]
Fixed in [``v0.4.0``](https://github.com/ofekashery/vertical-stack-in-card/releases/tag/v0.4.0) 🎉

**Comment by jeremysherriff** [comment]
Nice work
